### ISSUE TITLE
Defining names for parameters on typed function tests

### ISSUE THREAD

**Issue by rhdunn** [qt4cg/qtspecs#1136]
When defining the type of a higher-order function parameter, you cannot currently specify the names of the parameters of that higher-order function.

Allowing this can be useful for variou reasons:
1. documenting the parameter names in the function signature -- this makes it clear looking at the function in an IDE, etc. what the parameters are;
2. making the specs clearer by referring to the parameters by name;
3. allowing a processor to provide better error messages by referring to the parameters names, e.g. when there is a type conversion error;
4. allowing a user to reference the parameter by name if we enable this to resolve named keyword argments (which is currently being discussed in #1114).

Thus, you could declare e.g. [index-where](https://qt4cg.org/specifications/xpath-functions-40/Overview.html#func-index-where) like this:
```xquery
declare function fn:index-where(
  $input as item()*,
  $predicate as function(
    $item as item(),
    $position as xs:integer
  ) as xs:boolean
) as xs:integer* {
  (: ... :)
};
```

**Comment by ChristianGruen** [comment]
Thanks for the separate proposal.

I’m not 100% sure I understand how 4. could look like in practice. Could you add a little example for it?

**Comment by rhdunn** [comment]
```
declare function fn:index-where(
  $input as item()*,
  $predicate as function(
    $item as item(),
    $position as xs:integer
  ) as xs:boolean
) as xs:integer* {
  for $it at $pos in $input
  where $predicate(item := $it, position := $pos)
  return $pos
};
```

Here, the processor resolves `$predicate` in the function body to the parameter declaration. That has a static type which defines the parameter names of the function bound to that parameter. That static type is used to resolve the named keywords.

**Comment by rhdunn** [comment]
It would make sense for `fn:count#1`, etc. to keep the parameter names as part of the typed function test. -- This would allow the use of keyword arguments in that case as I discussed in the other thread.

I think it makes sense for function coercion rules, subsumption rules, etc. to allow different names -- otherwise there would be incompatibilities with 3.1 -- and use the names in the nearest scope.

That is, given:
```
let $f := fn:count#1 return $f(input := (1, 2))
```
the `fn:count#1` expression resolves to the static type `function($input as item()*) as xs:integer` as you note. The variable `$f` inherits that inferred static type. And the dynamic call to `$f(...)` makes use of that static type.

And given:
```
let $f as function($values as item()*) as xs:integer := fn:count#1
return $f(values := (1, 2))
```
the variable `$f` provides a specific static type. And the dynamic call to `$f(...)` makes use of that specific static type.

**Comment by rhdunn** [comment]
@michaelhkay I'm happy to keep this constrained, such that your `$actual-predicate` example would not be able to use named parameters without the dynamic proposal from @ChristianGruen. -- This goes along the lines of your approach of introducing small, constrained features rather than trying to do everything at once.

There are two parts to this proposal:
1. supporting naming the parameters on a function test as a purely documentational feature -- i.e. where it does not affect the semantics of the function test;
2. extending the use of named keywords to the case where the static type of a variable/parameter is a function test with parameter names, or via a named function reference.

I don't expect this specific proposal to go any/much farther than that. Doing so would require the dynamic proposal that @ChristianGruen is putting forward.

**Comment by ChristianGruen** [comment]
> @ChristianGruen. -- This goes along the lines of your approach of introducing small, constrained features rather than trying to do everything at once.

An even smaller step (hopefully not too small) would be to introduce optional parameter names, but to utilize them only for documentation and error messages. This would address reason 1-3 of this proposal.

Closely related: #981.

**Comment by rhdunn** [comment]
I'm happy to treat 1-3 and 4 as two separate proposals.

**Comment by michaelhkay** [comment]
I do think we should avoid the "thin end of the wedge" effect of static type inferencing. I don't think it would be acceptable to allow parameter names on `$param(....)` but not on `$param[3](....)` or `$param()(....)` or `$record?callback(....)`.

**Comment by michaelhkay** [comment]
I agree that allowing names purely for documentation purposes seems a sensible idea.

It's hard to separate this, however, from the fact that according to the data model, a function item has parameters with names. Currently there is no way of determining what the names are, and I don't think there is anything in the language that would be any different if they didn't have names, but we should make a decision whether to retain the names and make use of them, or to drop them.

**Comment by ChristianGruen** [comment]
> I agree that allowing names purely for documentation purposes seems a sensible idea.
> 
> It's hard to separate this, however, from the fact that > we should make a decision whether to retain the names and make use of them, or to drop them.

Now that we introduce keyword parameters with version 4, I think it would be just consistent to keep the names as well.
